light/detector.go

```diff
@@ -0,0 +1,223 @@
+
+// detectDivergence is a second wall of defense for the light client and is used
+// only in the case of skipping verification which employs the trust level mechanism.
+//
+// It takes the target verified header and compares it with the headers of a set of
+// witness providers that the light client is connected to. If a conflicting header
+// is returned it verifies and examines the conflicting header against the verified
+// trace that was produced from the primary. If successful it produces two sets of evidence
+// and sends them to the opposite provider before halting.
+//
+// If there are no conflictinge headers, the light client deems the verified target header
+// trusted and saves it to the trusted store.
+func (c *Client) detectDivergence(primaryTrace []*types.LightBlock, now time.Time) error {
+	var (
+		headerMatched      bool
+		lastVerifiedHeader = primaryTrace[len(primaryTrace)-1].SignedHeader
+		witnessesToRemove  = make([]int, 0)
+	)
+	c.logger.Info("Running detector against trace", "endBlockHeight", lastVerifiedHeader.Height,
+		"endBlockHash", lastVerifiedHeader.Hash, "length", len(primaryTrace))
+
+	c.providerMutex.Lock()
+	defer c.providerMutex.Unlock()
+
+	if len(c.witnesses) == 0 {
+		return errNoWitnesses{}
+	}
+
+	// launch one goroutine per witness to retrieve the light block of the target height
+	// and compare it with the header from the primary
+	errc := make(chan error, len(c.witnesses))
+	for i, witness := range c.witnesses {
+		go c.compareNewHeaderWithWitness(errc, lastVerifiedHeader, witness, i)
+	}
+
+	// handle errors from the header comparisons as they come in
+	for i := 0; i < cap(errc); i++ {
+		err := <-errc
+
+		switch e := err.(type) {
+		case nil: // at least one header matched
+			headerMatched = true
+		case errConflictingHeaders:
+			// We have conflicting headers. This could possibly imply an attack on the light client.
+			// First we need to verify the witness's header using the same skipping verification and then we
+			// need to find the point that the headers diverge and examine this for any evidence of an attack.
+			//
+			// We combine these actions together, verifying the witnesses headers and outputting the trace
+			// which captures the bifurcation point and if successful provides the information to create
+			witnessTrace, primaryBlock, err := c.examineConflictingHeaderAgainstTrace(primaryTrace, e.Block.SignedHeader, e.Witness, now)
+			if err != nil {
+				c.logger.Info("Error validating witness's divergent header", "witness", e.Witness, "err", err)
+				witnessesToRemove = append(witnessesToRemove, e.Index)
+				continue
+			}
+			// We are suspecting that the primary is faulty, hence we hold the witness as the source of truth
+			// and generate evidence against the primary that we can send to the witness
+			ev := &types.LightClientAttackEvidence{
+				ConflictingBlock: witnessTrace[len(witnessTrace)-1],
```

Comment:
why `-1`?

Reply:
This was supposed to be the last block in the trace of the witness but this is actually the trusted block not the conflicting one. I have replaced it with the conflicting block now (`primaryBlock`)